### ISSUE TITLE
MenuComplete filtering duplicates when it shouldn't be

### ISSUE THREAD

**Issue by thegraffix** [PowerShell/PSReadLine#3896]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://www.powershellgallery.com/packages/PSReadLine)
- [X] Search the existing issues, especially the pinned issues.

### Exception report

```console
N/A
```

### Screenshot

Test files:
![explorer](https://github.com/PowerShell/PSReadLine/assets/57874058/38d196b1-76b2-4c11-9308-2eac252d1ed4)

<BR>How the test function behaves (correctly) in ISE when using CTRL+SPACE:
![ISE](https://github.com/PowerShell/PSReadLine/assets/57874058/cacebb8a-2a06-414c-bbb3-b1ecb0ac9711)

<BR>How it behaves (correctly) in VSCode:
![VSCode](https://github.com/PowerShell/PSReadLine/assets/57874058/b9608f60-1c20-48a9-8a37-898607d13f23)

<BR></BR>How it behaves in the console (with PSReadLine).  The duplicates have been eliminated when they shouldn't be; although the ListItemText is identical for the duplicate CompletionMatches, the objects represented by each CompletionResult are unique.
![pwsh Console](https://github.com/PowerShell/PSReadLine/assets/57874058/cc2231c9-18ff-4078-bb7b-1a0928e26d7d)

### Environment data

```console
PS Version: 7.4.0
PS HostName: ConsoleHost
PSReadLine Version: 2.3.4
PSReadLine EditMode: Windows
OS: 10.0.19041.1 (WinBuild.160101.0800)
BufferWidth: 210
BufferHeight: 9000
```

### Steps to reproduce

This is the test function I used.

Auto-complete (using either `TAB` or `CTRL+SPACE`) should only show files (and not folders) that start with "file" + a number between 2 and 9 (regardless of what comes after the number).

```
<# Get-ChildItem C:\test results:

file3.txt  <--- This is a directory
file1 - Copy.txt
file1.txt
file2 - Copy.txt
file2.txt
file4 - Copy.txt
file4.txt
[file].txt
#>

$gciParams = @{
    Path = Join-Path -Path 'C:\Test' -ChildPath '*'
    Filter = 'file*.*'
    File = $true
    Force = $true
}
<# (gci @gciParams).Name results:

file1 - Copy.txt
file1.txt
file2 - Copy.txt
file2.txt
file4 - Copy.txt
file4.txt
#>

$pattern = '.*[2-9].*'
$gciCommand = {Get-ChildItem @gciParams | Where-Object {$_.Name -match $pattern}}
<# (Invoke-Command $gciCommand).Name results:

file2 - Copy.txt
file2.txt
file4 - Copy.txt
file4.txt
#>

function Test-CompletionResult
{
    [CmdletBinding()]
    param (
        [Parameter(Mandatory, Position = 0, ParameterSetName = 'Num', ValueFromPipeline, ValueFromPipelineByPropertyName)]
        [ArgumentCompleter({
            param ($Command, $Parameter, $WordToComplete, $CommandAst, $FakeBoundParams)
            $objList = @()
            (Invoke-Command -ScriptBlock $gciCommand).foreach({
                $filePath = $_.FullName
                $fileNum  = ($_.Name | Select-String -Pattern '\d').Matches.Value
                $objList += [pscustomobject] @{
                    Path = $filePath
                    Num  = $fileNum
                }
            })
            ($objList | Where-Object {$_.Num -like "*$WordToComplete*"}).foreach({
                [System.Management.Automation.CompletionResult]::new($_.Num, $_.Num, 'ProviderItem', $_.Path)
            })
        })]        
        [int]$FileNum
    )
    process {Write-Output $FileNum}
}
```

### Expected behavior

- MenuComplete items shouldn't be filtered out (or recognized as "duplicates") when the ListItemText is the same yet the CompletionResults are different for each MenuComplete item (as the CompletionResult tooltip(s) in my example illustrate)
- Expected behavior should mirror how ISE and VSCode present CompletionResult items

### Actual behavior

The "duplicate" MenuComplete items are filtered out.

**Comment by thegraffix** [comment]
It appears this issue still persists in PSReadLine 2.3.5

Should I create a new issue or does it make more sense to re-open this one?
<BR></BR>

![2024-05-17 14 30 06](https://github.com/PowerShell/PSReadLine/assets/57874058/a58618c6-c998-48db-9a0a-e1e6be0d2921)

**Comment by daxian-dbw** [comment]
The fix was included in the [v2.4.0-beta0 release](https://github.com/PowerShell/PSReadLine/releases/tag/v2.4.0-beta0), not in v2.3.5, which is a v2.3 servicing release. Please upgrade to v2.4.0-beta0 and give it a try.
